src/main/java/edu/byu/cs/autograder/compile/ModuleIndependenceVerifier.java

```diff
@@ -0,0 +1,59 @@
+package edu.byu.cs.autograder.compile;
+
+import edu.byu.cs.autograder.GradingContext;
+import edu.byu.cs.autograder.GradingException;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ModuleIndependenceVerifier implements StudentCodeInteractor {
+    @Override
+    public void interact(GradingContext context, StudentCodeReader reader) throws GradingException {
+        Set<File> serverFiles = reader.filesMatching(".*server/src/main/java/.*").collect(Collectors.toSet());
+        Path serverPath = context.stageRepo().toPath().resolve("server/src/main/java");
+        Set<String> serverClasses = qualifiedNames(serverFiles, serverPath);
+
+        Set<File> clientFiles = reader.filesMatching(".*client/src/main/java/.*").collect(Collectors.toSet());
+        Path clientPath = context.stageRepo().toPath().resolve("client/src/main/java");
+        Set<String> clientClasses = qualifiedNames(clientFiles, clientPath);
+
+        try {
+            checkImports(context, reader, serverFiles, clientClasses);
+            checkImports(context, reader, clientFiles, serverClasses);
+        } catch (IOException e) {
+            throw new GradingException("Unable to verify module independence", e);
+        }
+    }
+
+    private Set<String> qualifiedNames(Set<File> files, Path relativeTo) {
+        Set<String> classes = new HashSet<>();
+        for(File file : files) {
+            String fileName = relativeTo.relativize(file.toPath()).toString();
+            if(!fileName.contains(".java")) continue;
+            classes.add(fileName.substring(0, fileName.indexOf('.')).replace('/', '.'));
```

Comment:
It is really nice that Java requires this so we can read the class names out of the files!

What if a student has a `public` class nested inside another class?
1. If they used it properly within it's own module, would it be OK?
2. If they accessed it improperly from the wrong module, would this detect it?

If it passes (1) but fails (2), I think that's ok. I just can't fail (1).

Reply:
1. Yes. This is only detecting what they do wrong, not what they do correctly.
2. I suppose not. For that matter, I wouldn't detect static imports either. Do you think that's a big deal?

Reply:
Ok, We should be detecting static imports now. 1. is still always yes. 2 I'm unsure of atm

Reply:
Looks like java doesn't allow you to import nested classes without importing the base containing class. Further more, only the nested class import is necessary. So the answers to both questions are yes

Reply:
Great! Nice work testing that and determine that it always works!